observer/probers/tls/tls.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// Check if the root of the leaf's certificate is the one we expect.
+func (p TLSProbe) isRootExpected(rootO, rootCN string) bool {
+	if p.rootOrg == rootO && p.rootCN == rootCN {
+		return true
+	}
+	return false
+}
+
+// Export time to (or since) expiration (in seconds) and reason (with
+// corresponding badOutcome label) to Prometheus.
+func (p TLSProbe) exportMetrics(expiryTime time.Duration, reason reason) {
+	p.certExpiry.WithLabelValues(p.url).Set(expiryTime.Seconds())
+	p.certExpiry.WithLabelValues(p.url, badOutcomeError{reason}.Error()).Set(float64(reason))
+}
+
+// Probe performs the configured TLS protocol. Return true if both root AND
+// response are the expected values, otherwise false. Export time to cert expiry
+// and reason as Prometheus metrics.
+func (p TLSProbe) Probe(timeout time.Duration) (bool, time.Duration) {
+	start := time.Now()
+	conn, secure_err := tls.Dial("tcp", p.url+":443", &tls.Config{})
+	if secure_err != nil {
+		// TLS connection failed, so try insecure connection instead.
+		rootO, rootCN, timeUntilExpiry, err := p.getInsecureCertInfo()
+		// Insecure connection successful, but certificate chain is invalid.
+		if errors.Is(err, fmt.Errorf(reasonToString[issuerVerifyFailed])) {
+			p.exportMetrics(timeUntilExpiry, issuerVerifyFailed)
+			return false, time.Since(start)
+		}
+		// Insecure connection failed.
+		if err != nil {
+			p.exportMetrics(timeUntilExpiry, internalError)
+			return false, time.Since(start)
+		}
+		// Insecure connection successful and chain is valid, so check if
+		// certificate is and should be expired.
+		if p.response == "expired" {
+			certInvalidErr := x509.CertificateInvalidError{}
+			p.exportMetrics(timeUntilExpiry, none)
+			return (errors.As(secure_err, &certInvalidErr) && certInvalidErr.Reason == x509.Expired) && p.isRootExpected(rootO, rootCN), time.Since(start)
+		}
+		// None of the above cases. Secure TLS connection failed and certificate
+		// should not be expired, so return false.
+		p.exportMetrics(timeUntilExpiry, internalError)
+		return false, time.Since(start)
+	}
+	// Certificate has been validated. Now check if it's revoked.
+	defer conn.Close()
+	chains := conn.ConnectionState().VerifiedChains
+	leaf, issuer, rootCert := chains[0][0], chains[0][1], chains[0][len(chains[0])-1].Issuer
+	timeUntilExpiry := time.Until(leaf.NotAfter)
+	rootO, rootCN := rootCert.Organization[0], rootCert.CommonName
+
+	// Check OCSP to see if the certificate is valid, revoked or unknown.
+	ocspStatus, err := getOCSPStatus(leaf, issuer)
+	if err != nil {
+		p.exportMetrics(timeUntilExpiry, ocspError)
+		return false, time.Since(start)
+	}
+	if ocspStatus == ocsp.Unknown {
+		p.exportMetrics(timeUntilExpiry, ocspUnknown)
+		return false, time.Since(start)
+	}
+
+	p.exportMetrics(timeUntilExpiry, none)
+	return ((p.response == "revoked" && ocspStatus == ocsp.Revoked) || (p.response == "valid" && ocspStatus == ocsp.Good)) && (p.isRootExpected(rootO, rootCN)), time.Since(start)
```

Comment:
This return statement is rather complex. It could be simplified in a few ways:
1) Change `getOCSPStatus` to something like `checkOCSPStatus`, which takes the expected status as an input and returns whether or not the fetched status matches, rather than returning the fetched status itself.
2) Separate this into multiple separate returns. We can do the "isRootExpected" check as soon as we have the chains -- if the root isn't the expected root, let's return immediately, and not even bother with the OCSP check! So that can happen up on line 179ish.